airbyte-integrations/connectors/destination-iceberg-v2/src/main/kotlin/io/airbyte/integrations/destination/iceberg/v2/IcebergSuperTypeFinder.kt

```diff
@@ -0,0 +1,137 @@
+import org.apache.iceberg.types.TypeUtil
+import org.apache.iceberg.types.Types.*
+
+/**
+ * A utility class that determines a "supertype" given two Iceberg [Type]s.
+ *
+ * The "supertype" is a type to which both input types can safely be promoted without data loss. For
+ * instance, INT can be promoted to LONG, FLOAT can be promoted to DOUBLE, etc.
+ *
+ * @property icebergTypesComparator comparator used to verify deep type equality.
+ */
+@Singleton
+class IcebergSuperTypeFinder(private val icebergTypesComparator: IcebergTypesComparator) {
+
+    /**
+     * Returns a supertype for [existingType] and [incomingType] if one exists.
+     * - If they are deeply equal (according to [IcebergTypesComparator.typesAreEqual]), returns the
+     * [existingType] as-is.
+     * - Otherwise, attempts to combine them into a valid supertype.
+     * - Throws [IllegalArgumentException] if no valid supertype can be found.
+     */
+    fun findSuperType(existingType: Type, incomingType: Type, columnName: String): Type {
+        // If the two types are already deeply equal, return one of them (arbitrary).
+        if (icebergTypesComparator.typesAreEqual(incomingType, existingType)) {
+            return existingType
+        }
+        // Otherwise, attempt to combine them into a valid supertype.
+        return combineTypes(existingType, incomingType, columnName)
+    }
+
+    /**
+     * Combines two top-level [Type]s. If exactly one is primitive and the other is non-primitive,
+     * no supertype is possible => throws [IllegalArgumentException].
+     */
+    private fun combineTypes(existingType: Type, incomingType: Type, columnName: String): Type {
+        if (existingType.isPrimitiveType != incomingType.isPrimitiveType) {
+            throwIllegalTypeCombination(existingType, incomingType, columnName)
+        }
+
+        // Both are primitive
+        if (existingType.isPrimitiveType && incomingType.isPrimitiveType) {
+            return combinePrimitives(
+                existingType.asPrimitiveType(),
+                incomingType.asPrimitiveType(),
+                columnName
+            )
+        }
+
+        // Both are non-primitive => not currently supported
+        throwIllegalTypeCombination(existingType, incomingType, columnName)
+    }
+
+    /**
+     * Checks whether either type is unsupported or unmapped (e.g. BINARY, DECIMAL, FIXED, etc.).
+     *
+     * @throws IllegalArgumentException if either type is unsupported.
+     */
+    private fun validateTypeIds(typeId1: TypeID, typeId2: TypeID) {
+        val unsupportedTypeIds = setOf(BINARY, DECIMAL, FIXED, UUID, MAP, TIMESTAMP_NANO)
+        if (typeId1 in unsupportedTypeIds || typeId2 in unsupportedTypeIds) {
+            val badTypeId = if (typeId1 in unsupportedTypeIds) typeId1 else typeId2
+            throw IllegalArgumentException(
+                "Unsupported or unmapped Iceberg type: $badTypeId. Please implement handling if needed."
+            )
+        }
+    }
```

Comment:
I find this slightly easier to read:

```suggestion
    private fun validateTypeIds(typeId1: TypeID, typeId2: TypeID) {
        val unsupportedTypes = setOf(BINARY, DECIMAL, FIXED, UUID, MAP, TIMESTAMP_NANO)
        val providedTypes = listOf(typeId1, typeId2)
        val foundUnsupported = providedTypes.filter { it in unsupportedTypes }
        
        if (foundUnsupported.isNotEmpty()) {
            throw IllegalArgumentException(
                "Unsupported or unmapped Iceberg type(s): ${foundUnsupported.joinToString()}. Please implement handling if needed."
            )
        }
    }
```

Reply:
done